### PR TITLE
PERF: axis=1 reductions with EA dtypes

### LINKED ISSUE #52016: PERF: pd.BooleanDtype in row operations is 2000000 times slower
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [x] I have confirmed this issue exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [x] I have confirmed this issue exists on the main branch of pandas.

### Reproducible Example

``` python
import pandas as pd
import numpy as np

shape = 250_000, 100
mask = pd.DataFrame(np.random.randint(0, 1, size=shape))


np_mask = mask.astype(bool)
pd_mask = mask.astype(pd.BooleanDtype())

assert all(isinstance(dtype, pd.BooleanDtype) for dtype in pd_mask.dtypes)
assert all(isinstance(dtype, np.dtype) for dtype in np_mask.dtypes)
# column operations are not that much slower
%timeit pd_mask.any(axis=0) # 16.3 ms
%timeit np_mask.any(axis=0) # 5.86 ms
# using pandas.BooleanDtype back end for ROW operations is MUCH SLOWER
%timeit pd_mask.any(axis=1) # 14.1 s 
%timeit np_mask.any(axis=1) # 6.73 ms
```

```
16.3 ms ± 178 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
5.86 ms ± 467 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
14.1 s ± 329 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
6.73 ms ± 255 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

### Edit: additional context and unexpected behavior
```python
import pandas as pd
import numpy as np
import pyarrow as pa

# Columns WIND_SPEED and WIND_GUST might be any number between 0 and 100, while the QC codes are 0-3
df = pd.DataFrame(
    {
        "WIND_SPEED": np.random.randint(0, 100, size=(100,)),
        "WIND_SPEED_QC": np.random.randint(0, 3, size=(100,)),
        "WIND_GUST": np.random.randint(0, 100, size=(100,)),
        "WIND_GUST_QC": np.random.randint(0, 3, size=(100,)),
    }
# I've been looking into the pyarrow dtypes and encountered a unexpected behavior
).astype(pd.ArrowDtype(pa.uint8()))
# the equality comparison returns a pd.BooleanDtype rather than pd.ArrowDtype
mask = df[["WIND_SPEED_QC", "WIND_GUST_QC"]].__ge__(1)
# this is not expected
assert all(isinstance(x, pd.BooleanDtype) for x in mask.dtypes)
# pd.BooleanDtype
%timeit mask.any(axis=1) # 5.35 ms
# pd.ArrowDtype
%timeit mask.astype(pd.ArrowDtype(pa.bool_())).any(axis=1) # 7.24 ms 
# np.bool_
%timeit mask.astype(bool).any(axis=1) # 197 µs
```
The pd.BooleanDtype is faster than the pd.ArrowDtype which makes sense, but if the backend is going to change
it would make sense to use the np.bool_ dtype.
```
5.35 ms ± 377 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
7.24 ms ± 278 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
197 µs ± 7.6 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 1a2e300170efc08cb509a0b4ff6248f8d55ae777
python           : 3.10.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.15.90.1-microsoft-standard-WSL2
Version          : #1 SMP Fri Jan 27 02:56:13 UTC 2023
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : C.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 2.0.0rc0
numpy            : 1.24.2
pytz             : 2022.7.1
dateutil         : 2.8.2
setuptools       : 59.6.0
pip              : 22.0.2
Cython           : None
pytest           : 7.2.2
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : 8.11.0
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : 2023.3.0
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 11.0.0
pyreadstat       : None
pyxlsb           : None
s3fs             : 2023.3.0
scipy            : None
snappy           : None
sqlalchemy       : 2.0.4
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : None
qtpy             : None
pyqt5            : None

</details>

### Prior Performance

_No response_

### LINKED ISSUE #54389: PERF: `DataFrame.all(axis="columns")` orders of magnitude slower for `bool[pyarrow]`
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this issue exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this issue exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
import numpy as np
from functools import reduce
import operator

data = np.random.randn(10_000, 10) > 0.5  # larger won't even finish
df_numpy = pd.DataFrame(data, dtype=bool)
df_arrow = df_numpy.astype("bool[pyarrow]")

%timeit df_numpy.all(axis="index")     # 208 µs ± 813 µs
%timeit df_numpy.all(axis="columns")   # 294 µs ± 3.4 µs

%timeit df_arrow.all(axis="index")     # 987 µs ± 3.93 µs
%timeit df_arrow.all(axis="columns")   # 812 ms ± 7.5 ms  (!!!)

%timeit reduce(operator.__and__, (s for _, s in df_arrow.items()))  # 421 µs ± 7.65 µs
```

Larger examples won't even run, so presumably it falls back to some pure python implementation.

### Installed Versions

<details>

```
INSTALLED VERSIONS
------------------
commit           : 232d84e9a5add831c753e67f1c88863442dc3c92
python           : 3.10.11.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.19.0-50-generic
Version          : #50-Ubuntu SMP PREEMPT_DYNAMIC Mon Jul 10 18:24:29 UTC 2023
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 2.1.0.dev0+1375.g232d84e9a5
numpy            : 1.24.4
pytz             : 2023.3
dateutil         : 2.8.2
setuptools       : 65.5.0
pip              : 23.2.1
Cython           : 0.29.33
pytest           : 7.4.0
hypothesis       : 6.82.0
sphinx           : 6.2.1
blosc            : 1.11.1
feather          : None
xlsxwriter       : 3.1.2
lxml.etree       : 4.9.3
html5lib         : 1.1
pymysql          : 1.4.6
psycopg2         : 2.9.6
jinja2           : 3.1.2
IPython          : 8.14.0
pandas_datareader: None
bs4              : 4.12.2
bottleneck       : 1.3.7
brotli           : 
fastparquet      : 2023.7.0
fsspec           : 2023.6.0
gcsfs            : 2023.6.0
matplotlib       : 3.7.2
numba            : 0.57.1
numexpr          : 2.8.4
odfpy            : None
openpyxl         : 3.1.2
pandas_gbq       : None
pyarrow          : 12.0.1
pyreadstat       : 1.2.2
pyxlsb           : 1.0.10
s3fs             : 2023.6.0
scipy            : 1.11.1
snappy           : 
sqlalchemy       : 2.0.19
tables           : 3.8.0
tabulate         : 0.9.0
xarray           : 2023.7.0
xlrd             : 2.0.1
zstandard        : 0.21.0
tzdata           : 2023.3
qtpy             : None
pyqt5            : None
```

</details>

### Prior Performance

_No response_

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] closes #52016 
- [x] closes #54389
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.1.0.rst` file if fixing a bug or adding a new feature.

cc @rhshadrach - this includes the test you wrote in #51923 (with a few edits) as it looks like axis=1 reductions with EA dtypes are not well tested.

This PR avoids special-casing the internal EA types (although special-casing might allow for further optimization). 

xref: #51474

Timings on a slow laptop:

```
import pandas as pd
import numpy as np

df = pd.DataFrame(np.random.randn(10000, 4), dtype="float64[pyarrow]")
%timeit df.sum(axis=1)

# 3.79 s ± 137 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)      -> main
# 1.35 s ± 40.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)     -> 2.0.3
# 250 ms ± 53 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)       -> 1.5.3
# 5.42 ms ± 306 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)  -> PR

df = pd.DataFrame(np.random.randn(10000, 4), dtype="Float64")
%timeit df.sum(axis=1)

# 1.85 s ± 44.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)     -> main
# 1.2 s ± 58.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)      -> 2.0.3
# 16.4 ms ± 745 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)  -> 1.5.3
# 5.15 ms ± 286 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)  -> PR
```

### CODE REVIEW

**Comment by rhshadrach** [comment]
Wow - this is quite clever! On my `AMD Ryzen 9 5950X` machine, I get

```
df = pd.DataFrame(np.random.randn(10000, 4), dtype="float64[pyarrow]")
%timeit df.sum(axis=1)
698 ms ± 2.82 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)  <-- main
929 µs ± 4.33 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)  <-- PR

df = pd.DataFrame(np.random.randn(10000, 4), dtype="Float64")
%timeit df.sum(axis=1)
401 ms ± 1.45 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)  <-- main
1.16 ms ± 2.52 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)  <-- PR
```

Also, for wide inputs, there appears to not be a slowdown

```
df = pd.DataFrame(np.random.randn(4, 10000), dtype="float64[pyarrow]")
%timeit df.sum(axis=1)
29 ms ± 157 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)  <-- main
28.2 ms ± 440 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)  <-- PR

df = pd.DataFrame(np.random.randn(4, 10000), dtype="Float64")
%timeit df.sum(axis=1)
22.9 ms ± 48.3 µs per loop (mean ± std. dev. of 7 runs, 10 loops each) <-- main
22.4 ms ± 273 µs per loop (mean ± std. dev. of 7 runs, 10 loops each) <-- PR
```

**Comment by lukemanley** [comment]
> I also ran ASVs for `frame_methods` and `stat_ops`. `frame_methods` showed no change

isn't that showing a huge change, seconds -> milliseconds?

**Comment by rhshadrach** [comment]
Right - I wasn't sure where axis=1 ops might show up (haven't thoroughly looked through the benchmarks); these seemed to be the most likely places.

**Review by rhshadrach — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/masked.py
Comment (lukemanley) [reply pandas/core/arrays/masked.py:1153]: hmmm, maybe not needed. I may have been mistaken, I'm going to try removing it.

File: pandas/tests/frame/test_reductions.py
Comment (lukemanley) [reply pandas/tests/frame/test_reductions.py:1967]: added

File: pandas/core/groupby/groupby.py
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:1910]: Ah, I see the issue with my request above; this would make axis=1 fail for SparseArray whereas it didn't before. I would be okay opening up an issue.

Comment (lukemanley) [reply pandas/core/groupby/groupby.py:1910]: sure - by opening an issue, do you mean xfail for now as part of this PR? or open an issue and address that first before this?

Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:1910]: Leave this PR as-is; open an issue to cleanup after this is merged (before is okay too)

Comment (lukemanley) [reply pandas/core/groupby/groupby.py:1910]: ah - got it. thanks

File: pandas/core/frame.py
Comment (jbrockmendel) [inline pandas/core/frame.py:11096]: can you add comments on what is going on here.  i think i get it, but it could be non-obvious to someone who doesnt know e.g. groupby.agg

**Comment by lukemanley** [comment]
> how would this compare to using reduce_axis_1 approach? i think that would avoid copies/allocations. agreed this is clever, but if theres a way to avoid depending on the groupby code id prefer it

I might be missing something but my read of `_reduce_axis1` is that it would only provide a benefit for 2D arrays whereas this specifically targets EA (1D) arrays.

**Comment by lukemanley** [comment]
The additional copy required to construct the 1D array would be nice to avoid. I took a quick look at using `_reduce_axis1` for `sum` and it looks promising - comparable performance to this PR and avoids the copy. As @rhshadrach points out it would only be an option for a subset of reduction types. Any thoughts on what to do with the other reductions?
 
Somewhat surprising is that with this current PR, `axis=1` reductions seem to be faster than `axis=0` reductions for EA dtypes:

```
import pandas as pd
import numpy as np

data = np.random.randn(1000, 1000)

df = pd.DataFrame(data, dtype="Float64")
%timeit df.sum(axis=0)
%timeit df.sum(axis=1)
# 150 ms ± 5.45 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
# 62.7 ms ± 3.15 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

df = pd.DataFrame(data, dtype="float64[pyarrow]")
%timeit df.sum(axis=0)
%timeit df.sum(axis=1)
# 301 ms ± 38.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# 63.6 ms ± 4.49 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

**Comment by jbrockmendel** [comment]
> Any thoughts on what to do with the other reductions?

median comes to mind.  Are there others?

At some point for #53261 I'm hoping to implement chunked-friendly implementations of the groupby reductions.  I think that combined with the trick in this PR could get the best of both worlds avoiding the copy.

File: pandas/core/frame.py
Comment (jbrockmendel) [inline pandas/core/frame.py:11087]: IIUC kurt is excluded here bc GroupBy doesnt support it.  Can you comment to that effect

File: pandas/core/frame.py
Comment (lukemanley) [reply pandas/core/frame.py:11151]: updated with comments

File: pandas/core/frame.py
Comment (lukemanley) [reply pandas/core/frame.py:11136]: added a comment here

**Comment by rhshadrach** [comment]
> > Any thoughts on what to do with the other reductions?
> 
> median comes to mind. Are there others?

It looks like just nunique and quantile.

**Comment by rhshadrach** [comment]
> Somewhat surprising is that with this current PR, axis=1 reductions seem to be faster than axis=0 reductions for EA dtypes:

It appears as if axis=0 has more overhead; I don't believe the results hold up for size e.g. (10000, 10000).

**Comment by jbrockmendel** [comment]
> It appears as if axis=0 has more overhead; I don't believe the results hold up for size e.g. (10000, 10000).

That seems reasonable.  I'd be interested in seeing profiling results though!

**Comment by rhshadrach** [comment]
> import pandas as pd
> import numpy as np
> 
> data = np.random.randn(1000, 1000)
> 
> df = pd.DataFrame(data, dtype="Float64")
> %timeit df.sum(axis=0)
> %timeit df.sum(axis=1)
> # 150 ms ± 5.45 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
> # 62.7 ms ± 3.15 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
> 
> df = pd.DataFrame(data, dtype="float64[pyarrow]")
> %timeit df.sum(axis=0)
> %timeit df.sum(axis=1)
> # 301 ms ± 38.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
> # 63.6 ms ± 4.49 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

On my slower laptop now:

```
data = np.random.randn(10000, 10000)

df = pd.DataFrame(data, dtype="Float64")
%timeit df.sum(axis=0)
%timeit df.sum(axis=1)
# 1.02 s ± 5.07 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# 2.62 s ± 46.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

df = pd.DataFrame(data, dtype="float64[pyarrow]")
%timeit df.sum(axis=0)
%timeit df.sum(axis=1)
# 1.24 s ± 7.72 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# 2.55 s ± 5.26 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Review by rhshadrach — APPROVED** [review]
lgtm; cc @jbrockmendel

**Comment by rhshadrach** [comment]
I would be okay with saying this closes the last two of the three issues mentioned in the OP

File: pandas/core/frame.py
Comment (phofl) [inline pandas/core/frame.py:11151]: Can you do copy=False here for CoW?

**Comment by rhshadrach** [comment]
@lukemanley - there's a whatsnew conflict here

**Comment by rhshadrach** [comment]
Thanks @lukemanley - very nice!

**Comment by topper-123** [comment]
Yeah, very nice indeed 👍 .

**Comment by jbrockmendel** [comment]
might also close #28487?  and less likely but related #36123?

**Comment by lukemanley** [comment]
> might also close #28487? and less likely but related #36123?

doesn't look like it closes either of these

**Pull request merged**
